hw/ip/otbn/dv/otbnsim/sim/wsr.py

```diff
@@ -0,0 +1,131 @@
+        raise NotImplementedError()
+
+    def read_signed(self) -> int:
+        '''Get the stored value as a 256-bit signed value'''
+        uval = self.read_unsigned()
+        return uval - (1 << 256 if uval >> 255 else 0)
+
+    def write_signed(self, value: int) -> None:
+        '''Set the stored value as a 256-bit signed value'''
+        assert -(1 << 255) <= value < (1 << 255)
+        uval = (1 << 256) + value if value < 0 else value
+        self.write_unsigned(uval)
+
+    def commit(self) -> None:
+        '''Commit pending changes'''
+        return
+
+    def changes(self) -> List[TraceWSR]:
+        '''Return list of pending architectural changes'''
+        return []
+
+
+class DumbWSR(WSR):
+    '''Models a WSR without special behaviour'''
+    def __init__(self, name: str):
+        super().__init__(name)
+        self._value = 0
+        self._next_value = None  # type: Optional[int]
+
+    def read_unsigned(self) -> int:
+        return self._value
+
+    def write_unsigned(self, value: int) -> None:
+        assert 0 <= value < (1 << 256)
+        self._next_value = value
+
+    def commit(self) -> None:
+        if self._next_value is not None:
+            self._value = self._next_value
+        self._next_value = None
+
+    def changes(self) -> List[TraceWSR]:
+        return ([TraceWSR(self.name, self._next_value)]
+                if self._next_value is not None
+                else [])
+
+
+class RandWSR(WSR):
+    '''The magic RND WSR'''
+    def __init__(self, name: str, seed: int):
+        super().__init__(name)
+        # RND has its own generator, independent of the rest of the process, to
+        # make it easier to ensure things stay reproducible.
+        #
+        # TODO: We've got to work out how to make the RTL and ISS do the same
+        #       thing. Probably model the RTL's LFSR here, and allow seeding to
+        #       simulate the CSRNG/EDN? Again, it's going to be more helpful to
+        #       have a local generator either way.
+        self.generator = random.Random(seed)
+
+    def read_unsigned(self) -> int:
+        return self.generator.getrandbits(256)
+
+    def read_u32(self) -> int:
+        '''Read a 32-bit unsigned result'''
+        return self.generator.getrandbits(32)
+
+    def write_unsigned(self, value: int) -> None:
+        return
+
+
+class WSRFile:
+    '''A model of the WSR file'''
+    def __init__(self, seed: int):
+        self.MOD = DumbWSR('MOD')
+        self.RND = RandWSR('RND', seed)
+        self.ACC = DumbWSR('ACC')
+
+    def _wsr_for_idx(self, idx: int) -> WSR:
+        assert 0 <= idx <= 2
+        return {
+            0: self.MOD,
+            1: self.RND,
+            2: self.ACC
+        }[idx]
+
+    def read_at_idx(self, idx: int) -> int:
+        '''Read the WSR at idx as an unsigned 256-bit value'''
+        return self._wsr_for_idx(idx).read_unsigned()
+
+    def write_at_idx(self, idx: int, value: int) -> None:
```

Comment:
Can we make this functionality available so that we can write

```
state.wsrs[0x2] = value
```

Reply:
Sorry, I didn't notice this comment. As I wrote in the big main reply to this, I'd much prefer to keep things super-explicit in the actual implementation and add the layer of syntactic sugar as part of document extraction. We've already been caught out several times by the magic in the original ISS, and I really want to make something where you can just glance at the code and *know* what's going on. So no overloading `__setitem__` / `__getitem__` or other such cleverness.